src/librustc_target/abi/call/wasm32.rs

```diff
@@ -1,16 +1,38 @@
-use crate::abi::call::{FnType, ArgType};
+use crate::abi::call::{FnType, ArgType, Uniform};
+use crate::abi::{HasDataLayout, LayoutOf, TyLayout, TyLayoutMethods};
 
-fn classify_ret_ty<Ty>(ret: &mut ArgType<'_, Ty>) {
-    ret.extend_integer_width_to(32);
+fn classify_ret_ty<'a, Ty, C>(cx: &C, ret: &mut ArgType<'a, Ty>)
+    where Ty: TyLayoutMethods<'a, C> + Copy,
+          C: LayoutOf<Ty = Ty, TyLayout = TyLayout<'a, Ty>> + HasDataLayout
+{
+    if ret.layout.is_aggregate() {
+        if let Some(unit) = ret.layout.homogeneous_aggregate(cx).unit() {
+            let size = ret.layout.size;
+            if unit.size == size {
+                ret.cast_to(Uniform {
+                    unit,
+                    total: size
+                });
+                return;
+            }
+        }
+
+        ret.make_indirect();
+    }
 }
 
 fn classify_arg_ty<Ty>(arg: &mut ArgType<'_, Ty>) {
-    arg.extend_integer_width_to(32);
+    if arg.layout.is_aggregate() {
+        arg.make_indirect_byval();
```

Comment:
Huh, looks like `byval` in the default in Clang's `ABIArgInfo`: https://github.com/llvm/llvm-project/blob/1ce75045ebf130fb3cc08b5b6041aab095137a47/clang/include/clang/CodeGen/CGFunctionInfo.h#L172-L182

This is also applied to indirect returns, does it just do nothing there?

Reply:
Can you clarify this? I'm not sure what you expect or how it differs from what you observe.

More generally, I think you're right that this ABI has some issues and that I need to audit it against the upstream C ABI. I'm confused about why this wasn't caught in tests, though. I'll try to figure that out.

Reply:
According to the [LangRef](https://llvm.org/docs/LangRef.html#id1036) `byval` can only be used on parameters, so I assume any code that applies it to return arguments just ignores it.

Reply:
To be clear: I was surprised you went for `byval` but it appears to be correct, so I don't think any change is needed here.